# Download dataset form  this website

## https://www.kaggle.com/faressayah/stanford-open-policing-project?select=police_project.csv

## Description::

On a typical day in the United States, police officers make more than 50,000 traffic stops. Our team is gathering, analyzing, and releasing records from millions of traffic stops by law enforcement agencies across the country. Our goal is to help researchers, journalists, and policymakers investigate and improve interactions between police and the public.


## Importing libraries::

In [1]:
import pandas as pd
import numpy as np

# Use Pandas' read_csv function  open it as a DataFrame

In [152]:
df = pd.read_csv("police_project.csv")

# exploring
# df.columns
# df.describe()

RangeIndex(start=0, stop=91741, step=1)

# What does each row represent?

#### hint::
head : Return the first n rows. (By default return first 5 rows.)

In [42]:
df.head()

0.9651627952605705

# How to get the basic statistics of all the columns?

In [43]:
df.describe()

county_name  driver_age_raw    driver_age
count          0.0    86414.000000  86120.000000
mean           NaN     1970.491228     34.011333
std            NaN      110.914909     12.738564
min            NaN        0.000000     15.000000
25%            NaN     1967.000000     23.000000
50%            NaN     1980.000000     31.000000
75%            NaN     1987.000000     43.000000
max            NaN     8801.000000     99.000000

# How to check the shape of dataset?

In [46]:
df.shape # 91741 records 15 features

(91741, 15)

# Check the type of columns?

In [50]:
# df.info()         !!!!!!!!!!!!!!! WRONG !!!!!!!!!!!!!!!!!!


# check type of columns
df.dtypes

stop_date              object
stop_time              object
county_name           float64
driver_gender          object
driver_age_raw        float64
driver_age            float64
driver_race            object
violation_raw          object
violation              object
search_conducted         bool
search_type            object
stop_outcome           object
is_arrested            object
stop_duration          object
drugs_related_stop       bool
dtype: object

# Locating missing Values?
#### detecting missing values
#### calculates the sum of each column


In [52]:
df.isna().sum()

stop_date                 0
stop_time                 0
county_name           91741
driver_gender          5335
driver_age_raw         5327
driver_age             5621
driver_race            5333
violation_raw          5333
violation              5333
search_conducted          0
search_type           88545
stop_outcome           5333
is_arrested            5333
stop_duration          5333
drugs_related_stop        0
dtype: int64

# Dropping Column that only contains missing values.

In [64]:
# county is completely empty
df.drop('county_name', axis=1, inplace=True) # axis 1 = columns

# """  EXTRA  """
################################################################################
# search_type has a large amount of missing values
# calculating percentaage of missing values

# df['search_type'].isna().sum()/(df['search_type'].isna().sum() + df['search_type'].value_counts().sum())

# missing/ total => 96% values are missing
################################################################################

'  EXTRA  '

# Do the men or women speed more often?

In [67]:
df['violation'].unique() # array(['Speeding', 'Other', 'Equipment', 'Moving violation', nan,
                         # 'Registration/plates', 'Seat belt'], dtype=object)

# which gender speeds more often
# first create a filter then pass the filter to data frame as index

filt = df['violation'] == 'Speeding'

speeding_df = df[filt] # 48463 rows out of total 91741

# speeding_df is a subset of original dataframe having only speeding violation data

speeding_df['driver_gender'].value_counts()

M    32979
F    15482
Name: driver_gender, dtype: int64

## ANSWER

### more males do speeding violation than females (32979 males - 15482 females)



# Which year had the least number of stops?

In [82]:
# convert stop_date to date from object
# PAGE 181 BOOK - CHAPTER 11 BOOK

time = pd.to_datetime(df['stop_date'])

time.dt.year.value_counts()

2012    10970
2006    10639
2007     9476
2014     9228
2008     8752
2015     8599
2011     8126
2013     7924
2009     7908
2010     7561
2005     2558
Name: stop_date, dtype: int64

## ANSWER

### 2005 had the least number of stops (2558)


# Does gender affect who gets searched during a stop?

In [109]:
# filter out records where searching is True
filt = df['search_conducted']

df[filt]['driver_gender'].value_counts()

M    2725
F     471
Name: driver_gender, dtype: int64

## ANSWER
### males get searched much more than females so gender does affect searching 



# How does drug activity change by time of day?

In [128]:
filt = df['drugs_related_stop'] 

# first have to convert stop_time (object) into more suitable datatype for using value_counts() 

time_of_day = pd.to_datetime(df[filt]['stop_time'])
time_of_day.dt.hour.value_counts() ## EXTRA ## time_of_day.dt.hour.value_counts().SORT_INDEX() to sort by index,,,

0     103
23     92
1      60
13     50
22     47
2      46
14     44
9      43
10     42
11     41
15     31
19     30
20     27
16     22
18     22
3      20
21     20
12     16
17     16
7      15
8      14
6      10
4       3
5       1
Name: stop_time, dtype: int64

## ANSWER
###  most drug activity takes place at hour 0 (103 stops) then hour 23 (92 stops)
### 60 stops at hour 1, so most drug activites are taking place during night time

# Do most stops occur at night?

In [151]:
# converting stop_time to datetime and then filtering out night time
time_of_stops = pd.to_datetime(df['stop_time'])

# considering 6 pm to 6 am as night

# stop should be after 5 pm   OR   before 7 am

filt = (time_of_stops.dt.hour > 17) | (time_of_stops.dt.hour < 7) # 8pm-6am

df[filt].value_counts()

df[filt].shape # 36260 rows

# df[filt].shape[0] / df.shape[0] night stops / total stops      40 percent

0.3952431301163057

## ANSWER
### 40 percent stops are ocurring at night